### PR TITLE
ISO15693 sniffing codec

### PR DESCRIPTION
**Pull request by ceres-c**
This PR adds sniffing support for ISO15693 tags.

The VCD -> VICC part of this is basically the same as the ISO15693 emulation codec already merged.

VICC -> VCD sniffing is implemented in single subcarrier mode only. When VCD demodulation is done, a reference "noise" value is acquired from the `DEMOD` signal (PORTA Pin 7, yellow in the screenshot below) to get a reference value of the field when no modulation is ongoing. This value is then fed to the DAC channel 0, which is the negative input to the analog comparator channel 0 (ACACH0). Positive input is, again, PORTA Pin 7, so whenever a signal goes beyond the noise value, an AC interrupt occurs.

The analog comparator interrupt will increase the DAC output (the AC threshold) to a slightly higher value and enable interrupts to wait for the third pulse. When the third pulse is found, the threshold will be increased again with an even higher value, this time around sampled by the ADC from `DEMOD-READER` signal (PORTA Pin 2, pink below)

SDS00001

Using two different signals is convenient for ease of data acquisition, given the low speed of the Xmega ADC. `DEMOD-READER` has the advantage of being a non-return-to-zero signal, but still follows `DEMOD`'s shape. Blindly sampling `DEMOD` would likely yield a wrong threshold, given it might be sampled too close to bottom or top values. On the other hand, `DEMOD-READER`, after a sufficient number of pulses (3) gives a likely reasonable value without struggling with DSP.

When the first pause after the 24 SOC pulses is found, if the number of previous pulses is considered sufficient (more than 15), the codec is initialized to decode upcoming data. From now on, every 18,88 us an interrupt will store the first half-bit in a register, then after another 18,88us the second half will be stored in the same register. When 16 bit halves are found, data is decoded to a byte, which is then appended to the codec buffer.

The rest is basically boilerplate and small changes made to accommodate for the timers being used elsewhere in the codebase.

In my tests this codec proved to be fairly resistant to "dirty" situations like moving the tag or the reader in the field or suddenly removing them. For best performance, the chameleon must be kept between the tag and the reader maximizing antenna coupling: in my case the tag had to be slightly off-center by a couple of millimeters, but YMMV.

Attached a photo of my (admittedly ugly, but reliable) test setup. Yellow probe corresponds to yellow `DEMOD` signal above (MCU Pin 5 -> PORTA Pin 7) and green probe is `DEMOD-READER` (MCU Pin 7 -> PORTA Pin 2). Below the Chameleon, covered by masking tape, was my ST M24LR-DISCOVERY demo board.
IMG_0589

As it turns out, if you're going to keep micro probes attached to your MCU pins for an extended period of time, you're going to need support for the cables or you're going to break your pins like I did. Thanks @MrMoDDoM for lending me your Chameleon :)

### CODE REVIEW

File: Firmware/Chameleon-Mini/Codec/Codec.h [hunk 1]
```diff
@@ -166,9 +182,6 @@ INLINE void CodecInitCommon(void) {
     EVSYS.CH0MUX = CODEC_DEMOD_IN_EVMUX0;
     EVSYS.CH1MUX = CODEC_DEMOD_IN_EVMUX1;
 
-    EVSYS.CH2MUX = CODEC_DEMOD_IN_EVMUX0;
```
Comment (fptrs) [inline Firmware/Chameleon-Mini/Codec/Codec.h:169]: @ceres-c did you make sure that no other codec needs the initialization here?

File: Firmware/Chameleon-Mini/Codec/ISO15693.c [hunk 2]
```diff
@@ -628,7 +628,7 @@ void ISO15693CodecTask(void) {
 
         if (DemodByteCount > 0) {
             LogEntry(LOG_INFO_CODEC_RX_DATA, CodecBuffer, DemodByteCount);
-            LEDHook(LED_CODEC_RX, LED_PULSE);
+            //LEDHook(LED_CODEC_RX, LED_PULSE);
```
Comment (fptrs) [inline Firmware/Chameleon-Mini/Codec/ISO15693.c:631]: @ceres-c I think we still want the LED to blink here

File: Firmware/Chameleon-Mini/Makefile [hunk 3]
```diff
@@ -5,13 +5,13 @@
 SHELL = /bin/sh
 
 #Supported configurations
-CONFIG_SETTINGS  += -DCONFIG_MF_CLASSIC_MINI_4B_SUPPORT
+#CONFIG_SETTINGS  += -DCONFIG_MF_CLASSIC_MINI_4B_SUPPORT
```
Comment (fptrs) [inline Firmware/Chameleon-Mini/Makefile:8]: @ceres-c let us keep the config settings from the master

File: Firmware/Chameleon-Mini/Codec/ISO15693.c (hunk 2 above)
Comment (ceres-c) [reply Firmware/Chameleon-Mini/Codec/ISO15693.c:631]: You're right, I really didn't spot this one

File: Firmware/Chameleon-Mini/Codec/Codec.h (hunk 1 above)
Comment (ceres-c) [reply Firmware/Chameleon-Mini/Codec/Codec.h:169]: I've been checking the codebase and it seems to ma that no other codec is using event channel 2 apart from iso14/iso15 sniff, but they both set  up their own event routing.
I have to say I have NOT tested emulation, I based this change on static code analysis only

File: Firmware/Chameleon-Mini/Makefile (hunk 3 above)
Comment (ceres-c) [reply Firmware/Chameleon-Mini/Makefile:8]: Fair enough

File: Firmware/Chameleon-Mini/Codec/SniffISO15693.c [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+INLINE void CardSniffInit(void) {
+    /* Reinit state variables */
+    CodecBufferPtr = CodecBuffer2;
+    ByteCount = 0;
+
+    if (CodecBuffer[0] & ISO15693_REQ_SUBCARRIER_DUAL) {
+        /**
+         * No support for dual subcarrier, 0xBAAAAAAD in log to mark unsupported behavior and stop here
+         */
+        *(CodecBufferPtr++) = 0xBA;
```
Comment (fptrs) [inline Firmware/Chameleon-Mini/Codec/SniffISO15693.c:286]: @ceres-c I would prefer a warning with the LOG_INFO_GENERIC type e.g. LogEntry(LOG_INFO_GENERIC, "NO DUAL SC", 10)

File: Firmware/Chameleon-Mini/Codec/SniffISO15693.c [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+                    (SampleRegisterH & 0b00000011) == 0b00000000 ||
+                    (SampleRegisterH & 0b00001100) == 0b00001100 ||
+                    (SampleRegisterH & 0b00001100) == 0b00000000 ||
+                    (SampleRegisterH & 0b00110000) == 0b00110000 ||
+                    (SampleRegisterH & 0b00110000) == 0b00000000 ||
+                    (SampleRegisterH & 0b11000000) == 0b11000000 ||
+                    (SampleRegisterH & 0b11000000) == 0b00000000
+                ) {
+                    /**
+                     * Check for invalid data coding (11 or 00 bit-halfs).
```
Comment (fptrs) [inline Firmware/Chameleon-Mini/Codec/SniffISO15693.c:621]: @ceres-c LOG_INFO_GENERIC would also be a better fit here

File: Firmware/Chameleon-Mini/Codec/SniffISO15693.c (hunk 4 above)
Comment (ceres-c) [reply Firmware/Chameleon-Mini/Codec/SniffISO15693.c:286]: Agreed

File: Firmware/Chameleon-Mini/Codec/SniffISO15693.c (hunk 5 above)
Comment (ceres-c) [reply Firmware/Chameleon-Mini/Codec/SniffISO15693.c:621]: Ok, I guess, I liked better to write straight in the buffer, but that's fine by me.
I'll need a global "failure" variable to have the logs in the right order: card data and then info about corrupted reception. Writing straight into the buffer did not require a specific code path in CodecTask

**Comment by ceres-c** [comment]
Ported suggested fixes

File: Firmware/Chameleon-Mini/Codec/SniffISO15693.c [hunk 6]
```diff
@@ -0,0 +1,763 @@
...
+
+    if(Flags.CardDemodFinished) {
+        Flags.CardDemodFinished = 0;
+
+        DemodByteCount = CardByteCount;
+
+        if (DemodByteCount > 0) {
+            LogEntry(LOG_INFO_CODEC_SNI_CARD_DATA, CodecBuffer2, DemodByteCount);
+            LEDHook(LED_CODEC_RX, LED_PULSE);
+
```
Comment (cacke-r) [inline Firmware/Chameleon-Mini/Codec/SniffISO15693.c:753]: Shouldn't we have a call to 'ApplicationProcess' here ? 
Or should this be added, once someone writes an App ?

File: Firmware/Chameleon-Mini/Codec/SniffISO15693.c [hunk 7]
```diff
@@ -0,0 +1,763 @@
...
+}
+
+void SniffISO15693CodecTask(void) {
+    if (Flags.ReaderDemodFinished) {
+        Flags.ReaderDemodFinished = 0;
+
+        DemodByteCount = ReaderByteCount;
+
+        if (DemodByteCount > 0) {
+            LogEntry(LOG_INFO_CODEC_SNI_READER_DATA, CodecBuffer, DemodByteCount);
```
Comment (cacke-r) [inline Firmware/Chameleon-Mini/Codec/SniffISO15693.c:739]: Shouldn't we have a call to 'ApplicationProcess' here ? 
Or should this be added, once someone writes an App ?

File: Firmware/Chameleon-Mini/Codec/SniffISO15693.c (hunk 6 above)
Comment (fptrs) [reply Firmware/Chameleon-Mini/Codec/SniffISO15693.c:753]: Yes it should be called here. Someone who writes an application would expect it to be here already. @ceres-c can you add it.

File: Firmware/Chameleon-Mini/Codec/SniffISO15693.c (hunk 7 above)
Comment (fptrs) [reply Firmware/Chameleon-Mini/Codec/SniffISO15693.c:739]: @ceres-c same her

**Comment by ceres-c** [comment]
Changes made

**Comment by ceres-c** [comment]
Now that this is merged, I'd suggest updating the wiki as well :)
https://github.com/emsec/ChameleonMini/wiki/Supported-Cards-and--Codecs

**Comment by fptrs** [comment]
Done, thanks for the hint

**Pull request merged**